Modify rule S1481: provide education description (#1590)
SonarSource/rspec: rules/S1481/javascript/rule.adoc

// introduction section
// empty for this rule

// first tab (root_cause section)
== Why is this an issue?

// rationale
If a local variable or a local function is declared but not used, it is dead code and should be removed. Doing so will improve maintainability because developers will not wonder what the variable or function is used for.

// impact
=== What is the potential impact?

==== Dead code

> Unused variable or local function usually happens after some code change, so that some logic becomes not needed anymore. In that case such code becomes unused and never executed. 

Reviewer: Some minor suggestion, as the original phrasing seemed a bit strange to me. Feel free to disregard
```suggestion
An unused variable or local function usually occurs after some code change, making some logic not needed anymore. In that case, such code becomes unused and will never be executed.
```

> Also, if you are writing code for the front-end then every unused variable or function that remains in your codebase is just extra bytes you have to send over the wire to your users. Unused code bloats your codebase unnecessarily and impacts the performance of your application.

Reviewer: ```suggestion
Also, if you are writing code for the front-end, every unused variable or function remaining in your codebase is just extra bytes you have to send over the wire to your users. Unused code bloats your codebase unnecessarily and impacts the performance of your application.
```

==== Wrong logic

> It could happen that due to a bad copy-paste or autocompletion the wrong variable is used, while the right one is only declared. In that case the unused variable should be used instead of deleted from the codebase.

Reviewer: I think the second part of the sentence `In that case the unused variable should be used instead of deleted from the codebase.` would make more sense in the "How to fix it" section instead.

==== Memory leaks

> Finally unused functions can also cause memory leaks. For example, an unused function can create a closure over a variable that would otherwise be released to the garbage collector.

Reviewer: ```suggestion
Finally, unused functions can also cause memory leaks. For example, an unused function can create a closure over a variable that would otherwise be released to the garbage collector.
```

[source,javascript]
----
let theThing = null;
const replaceThing = function () {
  const originalThing = theThing;
  const unused = function () {
    if (originalThing) {
      console.log("hi");
    }
  };
  theThing = {
    longStr: new Array(1000000).join("*"),
    someMethod: function () {
      console.log(someMessage);
    },
  };
};
setInterval(replaceThing, 1000);
----

// second tab: How can I fix it? ("How to fix it?" on SonarCloud) (how_to_fix section)
== How to fix it?

> Usually the fix for this issue is straightforward, you just need to remove unused variable declaration, or its name from the declaration statement if it's declared along with other variables.

Reviewer 2: Adda. comma and an article

"Usually, " and  "the unused variable"

> ==== Noncompliant Code Example

Reviewer: the `validate_asciidoc` check is failing because it's expecting a h4 level header to be under a h3, whereas here it's under a h2. I don't know whether we should be more lax with the checks or actually moving these headers as h3

> // introduction section

Author: I will remove these comments before merge
